File: oneirodex/routes_apis/saved_filters.py
"""Named filters and the tree behind them (INSP-3, v11 H-I).

A member's own rows only. There is no admin view of someone else's filters and
no sharing surface: a saved filter says what a person is looking for, which is
theirs, and INSP-29's smart collections are the same row with a tile rather
than a second, more public thing.
"""

from __future__ import annotations

from flask import current_app
from flask_login import current_user, login_required
from sqlalchemy import func, select

from oneirodex import db
from oneirodex.models import Game, SavedFilter
from oneirodex.schemas.saved_filters import (
    FilterPreviewBody,
    SavedFilterCreateBody,
    SavedFilterUpdateBody,
)
from oneirodex.utils.api_response import api_error, api_ok
from oneirodex.utils.filter_tree import (
    FilterTreeError,
    compile_filter_tree,
    describe_fields,
    parse_filter_tree,
)
from oneirodex.utils.library_acl import apply_game_access_filters
from oneirodex.utils.validation import validate_body

from . import apis_bp


def _own(filter_id: int) -> SavedFilter | None:
    return (
        db.session.execute(
            select(SavedFilter).filter_by(id=filter_id, user_id=current_user.id)
        )
        .scalars()
        .first()
    )


def _bad_tree(exc: FilterTreeError):
    # The path is the point: a builder can highlight the offending row instead
    # of saying "invalid filter" over a form with fifteen of them.
    return api_error(str(exc), code='bad_request', detail={'path': exc.path})


@apis_bp.route('/filters/fields', methods=['GET'])
@login_required
def saved_filter_fields():
    """What a builder may offer. The single source is ``filter_tree.FIELDS``."""
    return api_ok({'fields': describe_fields()})


@apis_bp.route('/filters/saved', methods=['GET'])
@login_required
def list_saved_filters():
    rows = (
        db.session.execute(
            select(SavedFilter)
            .filter_by(user_id=current_user.id)
            .order_by(SavedFilter.name)
        )
        .scalars()
        .all()
    )
    return api_ok({'filters': [row.to_dict() for row in rows]})


@apis_bp.route('/filters/saved', methods=['POST'])
@login_required
@validate_body(SavedFilterCreateBody)
def create_saved_filter(body: SavedFilterCreateBody):
    try:
        tree = parse_filter_tree(body.tree)
    except FilterTreeError as exc:
        return _bad_tree(exc)

    existing = (
        db.session.execute(
            select(SavedFilter).filter_by(user_id=current_user.id, name=body.name)
        )
        .scalars()
        .first()
    )
    if existing:
        return api_error(f'You already have a filter called "{body.name}".', code='conflict')

    row = SavedFilter(
        user_id=current_user.id,
        name=body.name,
        tree=tree,
        is_collection=body.is_collection,
    )
    db.session.add(row)
    try:
        db.session.commit()
    except Exception as exc:  # noqa: BLE001
        db.session.rollback()
        current_app.logger.warning('saved filter create failed: %s', exc)
        return api_error("Couldn't save that filter.", code='internal')
    return api_ok({'filter': row.to_dict()}, status=201)


@apis_bp.route('/filters/saved/<int:filter_id>', methods=['PUT'])
@login_required
@validate_body(SavedFilterUpdateBody)
def update_saved_filter(filter_id: int, body: SavedFilterUpdateBody):
    row = _own(filter_id)
    if not row:
        return api_error('Filter not found', code='not_found')

    if body.tree is not None:
        try:
            row.tree = parse_filter_tree(body.tree)
        except FilterTreeError as exc:
            return _bad_tree(exc)
    if body.name is not None and body.name != row.name:
        clash = (
            db.session.execute(
                select(SavedFilter).filter_by(user_id=current_user.id, name=body.name)
            )
            .scalars()
            .first()
        )
        if clash:
            return api_error(f'You already have a filter called "{body.name}".', code='conflict')
        row.name = body.name
    if body.is_collection is not None:
        row.is_collection = body.is_collection

    try:
        db.session.commit()
    except Exception as exc:  # noqa: BLE001
        db.session.rollback()
        current_app.logger.warning('saved filter update failed for %s: %s', filter_id, exc)
        return api_error("Couldn't save that filter.", code='internal')
    return api_ok({'filter': row.to_dict()})


@apis_bp.route('/filters/saved/<int:filter_id>', methods=['DELETE'])
@login_required
def delete_saved_filter(filter_id: int):
    row = _own(filter_id)
    if not row:
        return api_error('Filter not found', code='not_found')
    db.session.delete(row)
    try:
        db.session.commit()
    except Exception as exc:  # noqa: BLE001
        db.session.rollback()
        current_app.logger.warning('saved filter delete failed for %s: %s', filter_id, exc)
        return api_error("Couldn't delete that filter.", code='internal')
    return api_ok({'deleted': filter_id})


@apis_bp.route('/filters/preview', methods=['POST'])
@login_required
@validate_body(FilterPreviewBody)
def preview_filter(body: FilterPreviewBody):
    """How many titles this tree matches, for the builder to show as it is built.

    Counted through the member's own access filters, so the number is the one
    they would get on Browse rather than a household total they cannot reach.
    A malformed tree comes back as a 400 naming the part that is wrong, which
    is also why there is no separate validate route: this one already answers
    that question, and it answers it with a count attached.
    """
    try:
        clause = compile_filter_tree(body.tree, user=current_user)
    except FilterTreeError as exc:
        return _bad_tree(exc)

    base = apply_game_access_filters(select(Game).filter(clause), current_user)
    count = db.session.execute(select(func.count()).select_from(base.subquery())).scalar_one()

    sample = []
    if body.limit:
        rows = db.session.execute(base.order_by(Game.name).limit(body.limit)).scalars().all()
        sample = [{'uuid': g.uuid, 'name': g.name} for g in rows]
    return api_ok({'count': int(count), 'sample': sample})

File: frontend/member-app/src/api/savedFilters.ts
/** Named filters and the nested tree behind them (INSP-3). */
import { deleteJson, getJson, postJson, putJson } from './client'

/** A group combines its children; a leaf names one field. */
export type FilterNode = FilterGroup | FilterLeaf

export interface FilterGroup {
  op: 'and' | 'or' | 'not'
  nodes: FilterNode[]
}

export interface FilterLeaf {
  field: string
  value: boolean | string | string[]
}

export function isGroup(node: FilterNode): node is FilterGroup {
  return (node as FilterGroup).op !== undefined
}

/**
 * What a builder may offer.
 *
 * Served from the backend's own field registry rather than duplicated here:
 * the tree can only name fields the chip row already offers, and a second copy
 * of that list in the SPA is a copy that goes stale the day a field is added.
 */
export interface FilterField {
  field: string
  kind: 'flag' | 'enum' | 'enum_list' | 'text'
  values?: string[]
}

export interface SavedFilter {
  id: number
  name: string
  tree: FilterNode
  is_collection: boolean
  created: string | null
  updated: string | null
}

export interface FilterPreview {
  count: number
  sample: { uuid: string; name: string }[]
}

/**
 * A tree the server will not run, with the part that is wrong.
 *
 * `path` is dotted from the root (`root.1.0`), which is what lets the builder
 * mark the offending row instead of putting "invalid filter" above a form with
 * fifteen of them.
 */
export class FilterTreeRejected extends Error {
  path: string

  constructor(message: string, path: string) {
    super(message)
    this.name = 'FilterTreeRejected'
    this.path = path
  }
}

function rethrowTreeError(error: unknown): never {
  // `errorFromBody` hangs the whole envelope off `.data`, so the path the
  // server named for the offending node arrives there rather than on the error.
  const detail = (error as { data?: { detail?: { path?: string } } })?.data?.detail
  if (detail && typeof detail.path === 'string') {
    throw new FilterTreeRejected((error as Error).message, detail.path)
  }
  throw error
}

export async function fetchFilterFields(signal?: AbortSignal): Promise<FilterField[]> {
  const data = await getJson('/api/filters/fields', { signal, label: '/api/filters/fields' })
  return Array.isArray(data.fields) ? data.fields : []
}

export async function fetchSavedFilters(signal?: AbortSignal): Promise<SavedFilter[]> {
  const data = await getJson('/api/filters/saved', { signal, label: '/api/filters/saved' })
  return Array.isArray(data.filters) ? data.filters : []
}

export async function createSavedFilter(
  name: string,
  tree: FilterNode,
  isCollection = false,
): Promise<SavedFilter> {
  try {
    const data = await postJson('/api/filters/saved', {
      name,
      tree,
      is_collection: isCollection,
    })
    return data.filter
  } catch (error) {
    rethrowTreeError(error)
  }
}

export async function updateSavedFilter(
  id: number,
  patch: { name?: string; tree?: FilterNode; is_collection?: boolean },
): Promise<SavedFilter> {
  try {
    const data = await putJson(`/api/filters/saved/${id}`, patch)
    return data.filter
  } catch (error) {
    rethrowTreeError(error)
  }
}

export async function deleteSavedFilter(id: number): Promise<void> {
  await deleteJson(`/api/filters/saved/${id}`)
}

/** How many titles this tree matches, counted through the member's own access. */
export async function previewFilter(tree: FilterNode, limit = 0): Promise<FilterPreview> {
  try {
    const data = await postJson('/api/filters/preview', { tree, limit })
    return { count: data.count ?? 0, sample: Array.isArray(data.sample) ? data.sample : [] }
  } catch (error) {
    rethrowTreeError(error)
  }
}
